dashboard/lib/services/child_account/grace_period_handler.rb

```diff
@@ -0,0 +1,45 @@
+module Services
+  module ChildAccount
+    class GracePeriodHandler < Services::Base
+      LockoutSchedulingError = Class.new(StandardError)
+
+      attr_reader :user
+
+      # @param user [User] the student account
+      def initialize(user:)
+        @user = user
+      end
+
+      def call
+        return false unless eligible?
+
+        user.transaction do
+          Services::ChildAccount.start_grace_period(user)
+
+          estimated_lockout_date = Policies::ChildAccount.lockout_date(user)
+          raise LockoutSchedulingError, 'Lockout date is not set' unless estimated_lockout_date
+
+          CAP::LockoutJob.set(wait_until: estimated_lockout_date).perform_later(user_id: user.id)
+        end
+
+        true
+      end
+
+      # Checks if the user can be transited to the grace period state.
+      private def eligible?
+        return false if Policies::ChildAccount.compliant?(user)
+        return false if Policies::ChildAccount::ComplianceState.locked_out?(user)
+        # If the user is already in the grace period, it cannot be started again.
+        return false if Policies::ChildAccount::ComplianceState.grace_period?(user)
```

Comment:
"It depends on how we intend to use these services.

`LockoutHandler` returns true for already locked-out users because we use it not only to lock out the user but also to identify their locked-out state: https://github.com/code-dot-org/code-dot-org/blob/38ff97f57d25cdd6ea6ee749d4e3c644ae019209/dashboard/app/controllers/application_controller.rb#L342-L343

However, in `GracePeriodHandler`, we only need to determine whether the user should be transitioned to the grace period. We could make it return true for consistency with `LockoutHandler`, but this isn't currently used.

Reply:
It depends on how we intend to use a service.

`LockoutHandler` returns true for already locked-out users because we use it not only to lock out the user but also to identify their locked-out state: https://github.com/code-dot-org/code-dot-org/blob/38ff97f57d25cdd6ea6ee749d4e3c644ae019209/dashboard/app/controllers/application_controller.rb#L342-L343

However, in `GracePeriodHandler`, we only need to determine whether the user should be transitioned to the grace period. We could make it return true for consistency with `LockoutHandler`, but this isn't currently used: https://github.com/code-dot-org/code-dot-org/blob/38ff97f57d25cdd6ea6ee749d4e3c644ae019209/dashboard/app/controllers/application_controller.rb#L339-L340

Reply:
I believe a more important question is: should we clear the "grace period" state in `GracePeriodHandler` if the user becomes CAP compliant during the grace period?

Reply:
https://github.com/code-dot-org/code-dot-org/pull/59241/commits/cfd1806231b86bd4cb302f143818ea3eed806a46

Reply:
> I believe a more important question is: should we clear the "grace period" state in GracePeriodHandler if the user becomes CAP compliant during the grace period?

That's a good catch. It's probably good to have it because it's a possible transition if we mess up the launch.